Editorial: Error Terminology (#966)
graphql/graphql-spec: spec/Section 7 -- Response.md

response. The service's response describes the result of executing the requested
operation if successful, and describes any errors raised during the request.

> A response may contain both a partial response as well as a list of all _field

Reviewer: Plural?

> error_ in the case that any field error was raised on a field and was replaced

Reviewer 2: ```suggestion
error_ in the case that any _field error_ was raised on a field and was replaced
```

I'm indifferent to this suggestion since it's already linked in the paragraph, feel free to dismiss

### Errors

The `errors` entry in the response is a non-empty list of errors raised during
the _request_, where each error is a map of data described by the error result
format below.

If present, the `errors` entry in the response must contain at least one error.
If no errors were raised during the request, the `errors` entry must not be
present in the result.

If the `data` entry in the response is not present, the `errors` entry must be
present. It must contain at least one _request error_ indicating why no data was
able to be returned.

If the `data` entry in the response is present (including if it is the value

> {null}), the `errors` entry must also be present if a _field error_ was raised
> during execution.

Reviewer 2: context:
```
If the `data` entry in the response is present (including if it is the value
```

If we use "if and only if" then it would also enforce that errors must not be present if there were no field errors; I think this adds clarity.

```suggestion
{null}), the `errors` entry must be present if and only if a _field error_ was
raised during execution.
```

We could also change "a" to "one or more":

```suggestion
{null}), the `errors` entry must be present if and only if one or more _field
error_ were raised during execution.
```

contains should indicate why no data was able to be returned.

If the `data` entry in the response is present (including if it is the value
{null}), the `errors` entry in the response may contain a list of all _field
error_ that were raised during execution. If any _field error_ was raised during

> execution, it should contain the `errors` entry.

Reviewer 2: Because we've gone from plural to singular, the "it" has become ambiguous (it could be referring to the "field error"); so let's clarify:

```suggestion
execution, the request should contain the `errors` entry.
```

Reviewer 3: I believe the second sentence (If any field error was raised... it SHOULD (?) contain ...) is a repetition of the first sentence (the errors entry in the response MAY(?) contain a list...)  - basically saying 'errors' contains errors raised. 
At the same time, there is slight contradiction - MAY vs SHOULD

Reviewer 3: I suggest to remove the second sentence and change MAY -> SHOULD in the first.

Author: Lots of confusing interchanging of must/should - let me try to reword this section further to clarify

**Request errors**

:: A _request error_ is an error raised during a _request_ but before execution
begins. This may occur due to a parse grammar or validation error in the
_Document_, an inability to determine which operation to execute, or invalid

> input values for variables.

Reviewer 2: Note `CreateSourceEventStream()` can also raise a request error:

> 5. If groupedFieldSet does not have exactly one entry, raise a request error.
>
> -- https://spec.graphql.org/draft/#CreateSourceEventStream()

There's also various places where there are `Assert:` that could raise an error (though if you've done validation they should not occur, except [when we don't handle that well](https://github.com/graphql/graphql-spec/pull/860)).

The phrase "before execution begins" is quite a fuzzy definition it turns out; here's a comment where I try and pick it apart a bit: https://github.com/graphql/graphql-js/pull/3640#issuecomment-1153711229

My current opinion is that "execution begins"

- for queries: when you enter [ExecuteQuery()](https://spec.graphql.org/draft/#ExecuteQuery())
- for mutations: when you enter [ExecuteMutation()](https://spec.graphql.org/draft/#ExecuteMutation())
- for subscriptions: when you enter [MapSourceToResponseEvent()](https://spec.graphql.org/draft/#MapSourceToResponseEvent())

Author: Time boundary isn't the best way to describe this. Maybe prose about this being "typical" could be helpful, but really the difference is that field errors go alongside partial results where request errors do not

Reviewer 2: That's kind of circular reasoning though; for example if an error is raised from the `subscribe()` resolver of a field (i.e. in step 10 of [CreateSourceEventStream()](https://spec.graphql.org/draft/#CreateSourceEventStream()) via the [ResolveFieldEventStream()](https://spec.graphql.org/draft/#ResolveFieldEventStream()) algorithm) then it's not clear from the spec whether that should be a request error (since it prevents setting up the subscription, so the request cannot continue) or a field error (since it originated from a field). And thus, it's not clear if it should resolve to `{data: null, errors: [...]}` or just `{errors: [...]}` - and thus not clear if it's a partial result or not.

In my opinion, the boundary is effectively that it's a request error if it happens before the first `ExecuteSelectionSet()`; the boundaries I defined above give this rule of thumb explicit positions in the algorithm that are passed once only for each request.

A request error is typically the fault of the requesting client.

If a request error is raised, execution does not begin and the `data` entry in
the response must not be present. The `errors` entry must include the error.

**Field errors**

> :: A _field error_ is raised during the execution of a particular field. This

Reviewer 2: For consistency, either both "request error" and "field error" should have "is **an error** raised" or neither of them should. I think both is preferable:

```suggestion
:: A _field error_ is an error raised during the execution of a particular field. This
```
